rllib/utils/checkpoints.py

```diff
@@ -29,15 +37,287 @@
+        if "checkpoint_version" not in metadata:
+            metadata["checkpoint_version"] = str(
+                CHECKPOINT_VERSION_LEARNER_AND_ENV_RUNNER
+            )
+        with open(path / self.METADATA_FILE_NAME, "w") as f:
+            json.dump(metadata, f)
+
+        # Write the class and constructor args information to disk.
+        with open(path / self.CLASS_AND_CTOR_ARGS_FILE_NAME, "w") as f:
+            json.dump(
+                {
+                    "class": type(self),
+                    "ctor_args_and_kwargs": self.ctor_args_and_kwargs(),
+                },
+                f,
+            )
+
+        # Get the entire state of this Checkpointable, or use provided `state`.
+        state = state or self.get_state()
+
+        # Write components of `self` that themselves are `Checkpointable`.
+        for component_name, component in self.get_checkpointable_components():
+            # If subcomponent's name is not in `state`, ignore it and don't write this
+            # subcomponent's state to disk.
+            if component_name not in state:
+                continue
+            component_state = state.pop(component_name)
+            # By providing the `state` arg, we make sure that the component does not
+            # have to call its own `get_state()` anymore, but uses what's provided here.
+            component.save_to_path(path / component_name, state=component_state)
+
+        # Write all the remaining state to disk.
+        # Write state (w/o policies) to disk.
+        with open(path / self.STATE_FILE_NAME, "wb") as f:
+            pickle.dump(state, f)
+
+        return str(path)
+
+    def restore_from_path(self, path: Union[str, pathlib.Path], **kwargs) -> None:
+        """Restores the state of the implementing class from the given path.
+
+        The given `path` should have the following structure and contain the following
+        files:
+
+        .. testcode::
+            :skipif: True
+
+            path/
+                [component1]/
+                    [component1 subcomponentA]/
+                        ...
+                    [component1 subcomponentB]/
+                        ...
+                [component2]/
+                        ...
+                [cls.STATE_FILE_NAME].pkl
+
+        Note that the self.METADATA_FILE_NAME file is not required to restore the state.
+
+        Args:
+            path: The path to load the implementing class' state from.
+            **kwargs: Forward compatibility kwargs.
+        """
+        path = pathlib.Path(path)
+
+        # Restore components of `self` that themselves are `Checkpointable`.
+        for component_name, component in self.get_checkpointable_components():
+            component_dir = path / component_name
+            # If subcomponent's dir is not in path, ignore it and don't restore this
+            # subcomponent's state from disk.
+            if not os.path.isdir(component_dir):
```

Comment:
Can we stick here to the `pathlib` module and use `component_dir.is_dir()`?

Reply:
Great call. Always get confused with os.path vs pathlib. :|